homeassistant/components/tplink/__init__.py

```diff
@@ -64,101 +66,197 @@
-            hass, device
-        )
+    discovered_devices = {
+        dr.format_mac(device.mac): device
+        for device in (await Discover.discover()).values()
+    }
+    hass.data[DOMAIN][DISCOVERED_DEVICES] = discovered_devices
 
-        try:
-            await coordinator.async_refresh()
-        except SmartDeviceException:
-            _LOGGER.warning(
-                "Device at '%s' not reachable during setup, retry not yet implemented",
-                device.host,
-            )
+    if legacy_entry:
+        async_migrate_legacy_entries(hass, config_entries_by_mac, legacy_entry)
+
+    if conf is not None:
+        async_migrate_yaml_entries(hass, conf)
+
+    if discovered_devices:
+        async_trigger_discovery(hass, discovered_devices)
+
+    return True
+
+
+async def async_migrate_entities_devices(
+    hass: HomeAssistant, legacy_entry_id: str, new_entry: ConfigEntry
+) -> None:
+    """Move entities and devices to the new config entry."""
+    entity_registry = er.async_get(hass)
+    tplink_reg_entities = er.async_entries_for_config_entry(
+        entity_registry, legacy_entry_id
+    )
+
+    for reg_entity in tplink_reg_entities:
+        # Only migrate entities with a mac address only
+        if len(reg_entity.unique_id) < MAC_ADDRESS_LEN:
             continue
+        if dr.format_mac(reg_entity.unique_id[:MAC_ADDRESS_LEN]) == new_entry.unique_id:
+            entity_registry._async_update_entity(  # pylint: disable=protected-access
+                reg_entity.entity_id, config_entry_id=new_entry.entry_id
+            )
+
+    device_registry = dr.async_get(hass)
+    tplink_dev_entities = dr.async_entries_for_config_entry(
+        device_registry, legacy_entry_id
+    )
+    for dev_entry in tplink_dev_entities:
+        for connection_type, value in dev_entry.connections:
+            if (
+                connection_type == dr.CONNECTION_NETWORK_MAC
+                and value == new_entry.unique_id
+            ):
+                device_registry._async_update_device(  # pylint: disable=protected-access
+                    dev_entry.id, add_config_entry_id=new_entry.entry_id
+                )
+
+    hass.config_entries.async_update_entry(
+        new_entry,
+        data={k: v for k, v in new_entry.data.items() if k != CONF_LEGACY_ENTRY_ID},
+    )
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up TPLink from a config entry."""
+    if not entry.unique_id or entry.unique_id == DOMAIN:
+        return True
+
+    if legacy_entry_id := entry.data.get(CONF_LEGACY_ENTRY_ID):
+        await async_migrate_entities_devices(hass, legacy_entry_id, entry)
 
+    protocol = TPLinkSmartHomeProtocol(entry.data[CONF_HOST])
+    try:
+        info = await protocol.query(Discover.DISCOVERY_QUERY)
+    except SmartDeviceException as ex:
+        raise ConfigEntryNotReady from ex
+
+    device_class = Discover._get_device_class(info)  # pylint: disable=protected-access
```

Comment:
@rytilahti I _think_ this is going to be the only other change needed (making this public)

Reply:
Maybe just use `Discover.discover_single(host)` here?

Reply:
I did that first, but then we get two calls to .update().  One in the discover_single call, and then one immediately after when we refresh the coordinator. I wanted to avoid two calls when we only need the one in the coordinator.

Reply:
Maybe I can make it work that way by setting the coordinator data manually

Reply:
done in 1405f63680e1c3568c77408bda9fcf582ca57c9b